### PR TITLE
Add ability to set arbitrary ironic configs

### PR DESCRIPTION
**Pull request by derekhiggins**
As more use cases are using this image and may need to tweak
various config options. Add the method to set arbitary options
rather then having to add a new option for each case.

Potential solution (the ironic part) to https://github.com/metal3-io/ironic-image/issues/143

Another alternative method might be to mount a partial ironic.conf to be merged into /etc/ironic/ironic.conf , I've leaned in the direction of the env variable as its more consistent
with that we currently have

Marking as WIP until we discuss whats best.

### CODE REVIEW

**Comment by dtantsur** [comment]
Are you aware of https://specs.openstack.org/openstack/oslo-specs/specs/rocky/config-from-environment.html? Maybe it can be reused?

**Comment by derekhiggins** [comment]
> Are you aware of https://specs.openstack.org/openstack/oslo-specs/specs/rocky/config-from-environment.html? Maybe it can be reused?

thanks, I wasn't aware of this, I tried it out and it works, the only down side I can see is that somebody debugging the image might look at ironic.conf and not realise that it doesn't represent
all the config options, so I propose we also log any env options to the file system so that they would
be seen. something like this in configure-ironic.sh
`env | grep "^OS_" > /etc/ironic/ironic.env  `

**Comment by dhellmann** [comment]
> > Are you aware of https://specs.openstack.org/openstack/oslo-specs/specs/rocky/config-from-environment.html? Maybe it can be reused?
> 
> thanks, I wasn't aware of this, I tried it out and it works, the only down side I can see is that somebody debugging the image might look at ironic.conf and not realise that it doesn't represent
> all the config options, so I propose we also log any env options to the file system so that they would
> be seen. something like this in configure-ironic.sh
> `env | grep "^OS_" > /etc/ironic/ironic.env`

oslo.config can be made to log all of the known configuration options and their values. Apps typically do that by invoking `log_opt_values()` on startup. Does Ironic do that?

**Comment by derekhiggins** [comment]
> > > Are you aware of https://specs.openstack.org/openstack/oslo-specs/specs/rocky/config-from-environment.html? Maybe it can be reused?
> > 
> > 
> > thanks, I wasn't aware of this, I tried it out and it works, the only down side I can see is that somebody debugging the image might look at ironic.conf and not realise that it doesn't represent
> > all the config options, so I propose we also log any env options to the file system so that they would
> > be seen. something like this in configure-ironic.sh
> > `env | grep "^OS_" > /etc/ironic/ironic.env`
> 
> oslo.config can be made to log all of the known configuration options and their values. Apps typically do that by invoking `log_opt_values()` on startup. Does Ironic do that?

Yes it does, but sometimes in a long running ironic container the head of the log may be gone, also I typically exec into the container and usually look around at the setup.

File: configure-ironic.sh [hunk 1]
```diff
@@ -43,5 +43,9 @@ endpoint_override = http://${IRONIC_URL_HOST}:5050
 endpoint_override = http://${IRONIC_URL_HOST}:6385
 EOF
 
+# oslo.config also supports Config Opts From Environment, log them
+echo '# Options set from Environment variables' | tee /etc/ironic/ironic.conf
+env | grep "^OS_" | tee -a /etc/ironic/ironic.conf
```
Comment (dhellmann) [inline configure-ironic.sh:48]: This is adding shell environment variables directly to the config file. Could that cause parse issues? Should they be added as comments?

Comment (derekhiggins) [reply configure-ironic.sh]: thanks, that was a typo, I meant to add them to a different file.

Comment (dhellmann) [reply configure-ironic.sh]: Ah, cool, I assumed you wanted everything in one place for debugging. A separate file seems even safer, though.

**Pull request merged**